// @google/gemini-cli/dist/src/ui/commands/aboutCommand.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { getCliVersion } from '../../utils/version.js';
import { CommandKind } from './types.js';
import process from 'node:process';
import { MessageType } from '../types.js';
export const aboutCommand = {
    name: 'about',
    description: 'show version info',
    kind: CommandKind.BUILT_IN,
    action: async (context) => {
        const osVersion = process.platform;
        let sandboxEnv = 'no sandbox';
        if (process.env['SANDBOX'] && process.env['SANDBOX'] !== 'sandbox-exec') {
            sandboxEnv = process.env['SANDBOX'];
        }
        else if (process.env['SANDBOX'] === 'sandbox-exec') {
            sandboxEnv = `sandbox-exec (${process.env['SEATBELT_PROFILE'] || 'unknown'})`;
        }
        const modelVersion = context.services.config?.getModel() || 'Unknown';
        const cliVersion = await getCliVersion();
        const selectedAuthType = context.services.settings.merged.security?.auth?.selectedType || '';
        const gcpProject = process.env['GOOGLE_CLOUD_PROJECT'] || '';
        const ideClient = (context.services.config?.getIdeMode() &&
            context.services.config?.getIdeClient()?.getDetectedIdeDisplayName()) ||
            '';
        const aboutItem = {
            type: MessageType.ABOUT,
            cliVersion,
            osVersion,
            sandboxEnv,
            modelVersion,
            selectedAuthType,
            gcpProject,
            ideClient,
        };
        context.ui.addItem(aboutItem, Date.now());
    },
};
